### PR TITLE
Expose `HostConfiguration` type to Compose through Treehouse

### PR DESCRIPTION
**Pull request by JakeWharton**
This will hold more in the future such as the locale, pixel density, etc. For now we propagate whether the host is in dark mode to get things started.

https://user-images.githubusercontent.com/66577/192565428-61c9db20-764c-40bd-b46b-cdacb343dab3.mov

Refs #355

### CODE REVIEW

File: redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/ZiplineTreehouseUi.kt [hunk 1]
```diff
@@ -17,13 +17,15 @@ package app.cash.redwood.treehouse
...
 
 /**
  * Adapt TreehouseComposition to conform the limitations of Zipline interfaces.
  *
  * Most callers shouldn't use this directly; instead use `TreehouseUi`.
  */
 public interface ZiplineTreehouseUi : ZiplineService {
-  public fun start(diffSink: DiffSinkService)
+  // TODO: pass StateFlow end-to-end
+  public fun start(diffSink: DiffSinkService, hostConfigurations: Flow<HostConfiguration>)
```
Comment (JakeWharton) [inline redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/ZiplineTreehouseUi.kt]: https://github.com/cashapp/zipline/issues/726

File: redwood-treehouse-composeui/src/main/kotlin/app/cash/redwood/treehouse/composeui/TreehouseComposeView.kt [hunk 2]
```diff
@@ -60,6 +66,12 @@ public class TreehouseComposeView<T : Any>(
 
   override val protocolDisplayRoot: DiffConsumingWidget<*> = ProtocolDisplayRoot(this)
 
+  private val mutableHostConfiguration =
+    MutableStateFlow(computeHostConfiguration(context.resources.configuration))
+
+  override val hostConfiguration: StateFlow<HostConfiguration>
+    get() = mutableHostConfiguration
```
Comment (JakeWharton) [inline redwood-treehouse-composeui/src/main/kotlin/app/cash/redwood/treehouse/composeui/TreehouseComposeView.kt:75]: Wants https://github.com/Kotlin/KEEP/issues/278

File: redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/HostConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package app.cash.redwood.treehouse
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+public data class HostConfiguration(
+  val darkMode: Boolean = false,
+) {
+  public companion object
```
Comment (swankjesse) [inline redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/HostConfiguration.kt:24]: weeeird

File: redwood-treehouse/src/jsMain/kotlin/app/cash/redwood/treehouse/HostConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    throw AssertionError("HostConfiguration was not provided!")
+  }
+
+/**
+ * Expose various configuration properties of the host.
+ *
+ * @see HostConfiguration
+ */
+@Suppress("unused") // Emulating a CompositionLocal.
+public val HostConfiguration.Companion.current: HostConfiguration
```
Comment (swankjesse) [inline redwood-treehouse/src/jsMain/kotlin/app/cash/redwood/treehouse/HostConfiguration.kt:42]: Ahhh there it is

File: redwood-treehouse/src/jsMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt [hunk 5]
```diff
@@ -37,37 +40,39 @@ public fun TreehouseUi.asZiplineTreehouseUi(
...
 
-    override fun start(diffSink: DiffSinkService) {
-      check(diffSinkToClose == null)
-      diffSinkToClose = diffSink
-      delegate.start(diffSink)
-      startSignal()
+    val (initialHostConfiguration, hostConfigurationFlow) = hostConfigurations
+    composition.setContent {
+      val hostConfiguration by hostConfigurationFlow.collectAsState(initialHostConfiguration)
+      CompositionLocalProvider(LocalHostConfiguration provides hostConfiguration) {
```
Comment (swankjesse) [inline redwood-treehouse/src/jsMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt:67]: infix functions so weird

**Review by swankjesse — APPROVED** [review]

**Comment by swankjesse** [comment]
Really nice

File: redwood-treehouse/src/jsMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt (hunk 5 above)
Comment (JakeWharton) [reply redwood-treehouse/src/jsMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt:67]: Agreed

**Pull request merged**